feat: support queries for elements with local names

pkozlowski-opensource opened the pull request:
This PR adds support for `@ViewChildren` queries for elements with local names (ex. `@ViewChildren('foo')` foos;  should match `<div #foo>`. This is a first, small PR to update data structures as this new functionality requires storage of local names (ex. `#foo`) for a given node. As soon as we got all the necessary data stored the actual querying is very straightforward with the current infrastructure in place - the most important change is in packages/core/src/render3/query.ts

This PR also changes format of selector storage for queries - since in the current Angular we can only query for types or names (one or more) there is no reason to create / support more complex data structures that are prepared for queries by CSS selectors.

For now I'm not getting `ElementRef` from the injector as there are couple of small details I would like to discuss before.

pkozlowski-opensource committed:
feat: support queries for elements with local names

pkozlowski-opensource commented on packages/core/src/render3/query.ts:
```diff
@@ -111,14 +107,24 @@ function add(predicate: QueryPredicate<any>| null, node: LNode) {
+        const selector = predicate.selector !;
+        for (var i = 0; i < selector.length; i++) {
+          if (selector[i] === staticData.localName) {
+            predicate.values.push(new ElementRef(node.native));
```
We should be getting `ElementRef` from the DI instead. Need to discuss the `injectElementRef()` and similar functions first.

mhevery replied:
Yes, we should use DI, What is the question? You will need something like `injectElementRef` but you will have to pass in node explicitly somehow into it.

mhevery commented on packages/core/src/render3/instructions.ts:
```diff
@@ -321,13 +321,15 @@ export function getOrCreateNodeInjector(): LNodeInjector {
  * @param index Index of the element in the data array
  * @param nameOrComponentDef Name of the DOM Node or `ComponentDef`.
  * @param attrs Statically bound set of attributes to be written into the DOM element on creation.
+ * @param localName A name under which a given element is exported.
```
why is this a separate parameter? Why don't we just encode it into the attributes like so:
```
['#localName': '#']
```

Or some variation on that? This would also simplify querying?

pkozlowski-opensource replied:
Hmm, but I would either had to assume that it is stored as a first attribute _and_ / _or_ test for name (is it starting with `#`?) OR iterate over the entire attributes array _and_ test for presence of `#`. 

IMO the current proposal is much simpler for covering functionality in the current Angular.

mhevery commented on packages/core/src/render3/query.ts:
```diff
@@ -33,16 +36,9 @@ export interface QueryPredicate<T> {
 
   /**
-   * If selector then contains the selector parts where:
-   * - even index:
```
why delete all of this text?

pkozlowski-opensource replied:
Because I've changed the proposed format for a selector. We can only query for names, we can't use CSS selectors in the current Angular. 

Supporting CSS selectors would require more complex structure (to support negation, commas etc.). I don't think we want to introduce this complexity now given that it is not present in the current Angular. And I would _love_ proper design doc if we want to add CSS selectors for queries.

mhevery commented on packages/core/src/render3/query.ts:
```diff
@@ -58,7 +54,7 @@ export class QueryState_ implements QueryState {
 
-  track<T>(queryList: IQueryList<T>, predicate: Type<T>|any[], descend?: boolean): void {
+  track<T>(queryList: IQueryList<T>, predicate: Type<T>|string[], descend?: boolean): void {
     // TODO(misko): This is not right. In case of inherited state, a calling track will incorrectly
```
can you take care of this comment as well? Or is it separate PR?

pkozlowski-opensource replied:
Will do in a separate PR with proper tests for this use-case. Generally speaking I will try to open small & focused PRs so those are easier to review, if you don't mind.

mhevery commented on packages/core/src/render3/instructions.ts:
```diff
@@ -535,10 +538,11 @@ export function elementProperty<T>(index: number, propName: string, value: T | N
   return {
     tagName,
     attrs,
+    localName,
```
`localName: localName`

mhevery commented on packages/core/src/render3/di.ts:
```diff
@@ -86,12 +86,12 @@ export function bloomFindPossibleInjector(injector: LNodeInjector, bloomBit: num
 
-export function injectElementRef(): IElementRef {
-  let di = getOrCreateNodeInjector();
+export function injectElementRef(node?: LElement | LContainer): IElementRef {
```
`export injectElementRef: () => IELementRef = injectElementRefForNode`;

pkozlowski-opensource commented:
@mhevery rebased and addressed all the comments, PTAL

Note: a diff is bigger than it should be but it seems like `gulp format` wasn't run on one of the previous commits and running it on this PR introduced some re-formatting in the unrelated parts of the code.

mhevery committed:
style: fix formatting errors
